### PR TITLE
Blob Refactor

### PR DESCRIPTION
**Pull request by jessemapel**
## Description

This PR completes RFC 6, the Blob refactor so that Classes uses Blobs to serialize themselves, but don't inherit from Blob. This is API breaking and several classes were removed because their functionality independent of Blob was too small. This also adds new gtests for several of the Blob classes.

There are several failing app tests, but they are due to slightly different sizes in the Blobs because of how the storage method was cleaned up. Updated test data is ready to be checked in one this is merged.

## Related Issue

#4082 

## Motivation and Context

This makes it much easier to serial new objects to a cube because they do not need to inherit an entire new class with members, particularly for classes that already have a parent class. This also reduces the code base by a good amount and reduces the amount of functionality that must be tested on many classes.

## How Has This Been Tested?

All Blob classes were tested by hand and with automated tests. In particular, the CubeStretch/StretchBlob changes were hand tested and verified.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation change (update to the documentation; no code change)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have read and agree to abide by the [Code of Conduct](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/Code-Of-Conduct.md)
- [x] I have read the [**CONTRIBUTING**](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md) document.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] I have added myself to the [.zenodo.json](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/.zenodo.json) document.
- [x] I have added any user impacting changes to the [CHANGELOG.md](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CHANGELOG.md) document.

## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [x] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

### CODE REVIEW

File: isis/src/qisis/apps/qtie/QtieTool.cpp [hunk 1]
```diff
@@ -916,18 +916,18 @@ namespace Isis {
...
-    catch (IException &e) {
-      QString message = "Could not read cube history, "
-                        "will not update history.\n";
-      QString errors = e.toString();
-      message += errors;
-      QMessageBox::warning((QWidget *)parent(), "Warning", message);
-      return;
-    }
+    History h = p_matchCube->readHistory("IsisCube");
+    // try {
```
Comment (AustinSanders) [inline isis/src/qisis/apps/qtie/QtieTool.cpp]: Is this block something that can be removed?

**Review by AustinSanders — DISMISSED** [review]
Other than history comments, I didn't see anything that really needs to be changed.

File: isis/src/qisis/apps/qtie/QtieTool.cpp (hunk 1 above)
Comment (acpaquette) [reply isis/src/qisis/apps/qtie/QtieTool.cpp]: Probably? I commented this out as I didn't know how to add the functionality back in. I think checking the potential read in history for a number of objects > 0 would probably work?

**Review by acpaquette — DISMISSED** [review]
Overall looks good! There is one failing unit test, isis_unit_test_GisTopology. It's something that I think I or Jesse could take care of

File: isis/src/base/apps/csminit/csminit.cpp [hunk 2]
```diff
@@ -465,8 +465,11 @@ namespace Isis {
         cube->write(originalCameraStatistics);
       }
 
-      if (originalFootprint.Size() != 0) {
-        cube->write(originalFootprint);
+
+      if (originalFootprint.Polys() != NULL){
+        if (originalFootprint.Polys()->getNumGeometries() != 0) {
```
Comment (jessemapel) [inline isis/src/base/apps/csminit/csminit.cpp]: This should just be an AND instead of two IFs

File: isis/src/base/objs/Blob/Blob.cpp [hunk 3]
```diff
@@ -501,10 +538,14 @@ namespace Isis {
     }
   }
 
+  char *Blob::getBuffer() {
```
Comment (jessemapel) [inline isis/src/base/objs/Blob/Blob.cpp:546]: needs a doc string

File: isis/src/base/objs/OriginalXmlLabel/OriginalXmlLabel.cpp [hunk 4]
```diff
@@ -138,7 +127,7 @@ namespace Isis {
    *
    * @return @b QDomDocument The parsed original label
    */
-  const QDomDocument &OriginalXmlLabel::ReturnLabels() const {
+  const QDomDocument &OriginalXmlLabel::ReturnLabels(){
```
Comment (jessemapel) [inline isis/src/base/objs/OriginalXmlLabel/OriginalXmlLabel.cpp]: This can stay const

File: isis/src/qisis/apps/qtie/QtieTool.cpp (hunk 1 above)
Comment (jessemapel) [reply isis/src/qisis/apps/qtie/QtieTool.cpp]: I think this can go

File: isis/src/qisis/apps/qtie/QtieTool.cpp [hunk 5]
```diff
@@ -916,18 +916,18 @@ namespace Isis {
...
-    }
-    catch (IException &e) {
-      QString message = "Could not read cube history, "
-                        "will not update history.\n";
-      QString errors = e.toString();
-      message += errors;
-      QMessageBox::warning((QWidget *)parent(), "Warning", message);
-      return;
-    }
+    History h = p_matchCube->readHistory("IsisCube");
```
Comment (jessemapel) [inline isis/src/qisis/apps/qtie/QtieTool.cpp]: We don't need to specify the name here, the default will handle this for us

**Comment by jessemapel** [comment]
CI is good, the new app test failures are all expected and we have test data ready to check in. @acpaquette  @AustinSanders just need a re-approval on this to merge.

**Review by AustinSanders — APPROVED** [review]

**Pull request merged**